assets/js/components/Link.js

```diff
@@ -21,52 +21,118 @@
+
+const BUTTON = 'BUTTON';
+const BUTTON_DISABLED = 'BUTTON_DISABLED';
+const EXTERNAL_LINK = 'EXTERNAL_LINK';
+const LINK = 'LINK';
+const ROUTER_LINK = 'ROUTER_LINK';
+
 function Link( {
-	href,
+	'aria-label': ariaLabelProp,
+	arrow,
+	back,
+	caps,
 	children,
 	className,
-	arrow,
+	danger,
+	disabled,
 	external,
 	hideExternalIndicator,
+	href,
+	inherit,
 	inverse,
-	back,
+	onClick,
 	small,
-	inherit,
-	caps,
-	danger,
-	disabled,
-	'aria-label': ariaLabel,
-	...extraProps
+	to,
+	...otherProps
 } ) {
-	// Note: the disabled attribute does not alter behavior of anchor tags,
-	// so if disabled we force it to be a button.
-	const isAnchor = href && ! disabled;
-	const SemanticLink = isAnchor ? 'a' : 'button';
-
-	const getAriaLabel = () => {
-		let label = ariaLabel;
-
-		if ( ! external ) {
-			return label;
-		} else if ( typeof children === 'string' ) {
-			label = label || children;
+	const hasStringAsChild = typeof children === 'string';
+
+	// Append `aria-label` with text if label exists, otherwise return text.
+	const getLabelWithText = ( text ) => {
+		const shouldUseChildrenAsLabel = hasStringAsChild && ! ariaLabelProp;
+		const ariaLabel = shouldUseChildrenAsLabel ? children : ariaLabelProp;
+
+		return ariaLabel
+			? `${ ariaLabel } ${ text }`
+			: text;
+	};
+
+	// Do not create `aria-label` value if it would be identical to `children`,
+	// redundant label (bad a11y).
+	const getNonIdenticalLabel = () => {
+		const childIsIdenticalValue = hasStringAsChild && children === ariaLabelProp;
+
+		return childIsIdenticalValue ? undefined : ariaLabelProp;
+	};
+
+	const getType = () => {
+		// Force button element if `onClick` prop is passed.
+		if ( onClick ) {
+			return BUTTON;
+		}
+
+		// Disabled attribute does not alter behavior of anchors or links.
+		if ( disabled ) {
+			return BUTTON_DISABLED;
+		}
+
+		// Only `RouterLink` uses the `to` prop.
+		if ( to ) {
+			return ROUTER_LINK;
 		}
 
-		const newTabText = _x( '(opens in a new tab)', 'screen reader text', 'google-site-kit' );
-		if ( label ) {
-			return `${ label } ${ newTabText }`;
+		// The external prop means this is an external link, which will also output
+		// an `<a>` tag.
+		if ( external ) {
+			return EXTERNAL_LINK;
 		}
-		return newTabText;
+
+		// A regular `<a>` tag without external indicators.
+		return LINK;
 	};
 
+	const getLinkComponent = ( type ) => {
+		if ( type === BUTTON || type === BUTTON_DISABLED ) {
+			return 'button';
+		}
+
+		if ( type === ROUTER_LINK ) {
+			return RouterLink;
+		}
+
+		return 'a';
+	};
+
+	const getAriaLabel = ( type ) => {
+		if ( type === EXTERNAL_LINK ) {
+			return getLabelWithText( ARIA_TEXT_EXTERNAL );
+		}
+
+		if ( type === BUTTON_DISABLED ) {
+			return getLabelWithText( ARIA_TEXT_DISABLED );
+		}
+
+		return getNonIdenticalLabel();
+	};
+
+	const type = getType();
+	const LinkComponent = getLinkComponent( type );
+	const ariaLabel = getAriaLabel( type );
+
 	return (
-		<SemanticLink
+		<LinkComponent
```

Comment:
There's quite a few changes here and it seems like the component is getting more complex. Are there any tests we should add to `Link.test.js`?

Reply:
Looking at it further, it's mostly a refactoring of what was here before, but I've added tests for the component type and a few for the ARIA label behaviour.